enhancements/baremetal/continuous-hw-introspection.md

```diff
@@ -0,0 +1,135 @@
+---
+title: Continuous hardware introspecton of Baremetal IPI hosts
+authors:
+  - "@sadasu"
+reviewers:
+  - "@dhellmann"
+  - "@enxebre"
+  - "@hardys"
+  - "@stbenjam"
+  - "@juliakreger"
+approvers:
+  - "@dhellmann"
+  - "@enxebre"
+creation-date: 2020-02-18
+last-updated: 2020-02-17
+status: provisional|implementable|implemented|deferred|rejected|withdrawn|replaced
+---
+
+# Continuous hardware introspection of Baremetal IPI hosts
+
+This proposal talks about how performing continuous hardware introspection on all master
+and worker nodes in a Baremetal IPI deployment would get us important data about baremetal
+hosts which can be used to monitor changes in hardware capabilities of the server.
+
+This enhancement will not result in a new CRD or API to be written but will detail
+data that would become availabe in the Status field of the BareMetalHost CRD.
+
+## Release Signoff Checklist
+
+- [X] Enhancement is `implementable`
+- [X] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+Continuous hardware inspection on all the master and worker nodes that are a part
+of a Baremetal IPI deployment, allow the user to detect changes in server hardware
+and become aware of any failures seen on the servers. This information can be used
+to report current status of hardware to the user and/or glean metrics about the
+health of the deployment.
+
+At the end of 4.4, a Baremetal IPI deployment does not have continuous hardware introspection
+running on its master or worker nodes. The underlying provisioning service already has the
+capability to perform hardware inspection on each of the servers periodically. The
+baremetal-operator (BMO) then reads this information and updates the "status" field of the
+BareMetalHost CRD. Please refer to [1] for the latest BaremetalHost CRD.
+
+## Motivation
+
+This enhancement will allow the user to get continuous updates on all the baremetal
+servers in their Metal3 deployment providing useful information about the hardware
+capabilities of the servers.
+
+### Goals
+
+This enhancement will allow the baremetal host CR associated with each baremetal
+server in a metal3 deployment to contain up-to-date informtaion about its hardware and
+operational status of a server.
+
+### Non-Goals
+
+This proposal does not include any changes to the BareMetalHost CRD.
+
+## Proposal
+
+This enhancement proposes to add a new DaemonSet started by the machine-api-operator
+only for the BareMetal platform. This DaemonSet would run a new container image
+(ironic-hardware-inventory-recorder) on all master and worker nodes. With the new
```

Comment:
is there any scenario where we would want the machines to opt-out via e.g particular nodeSelector?

Reply:
The goal of this enhancement is to just focus on the work needed to enable collecting information for all master an worker nodes. We haven't envisioned this introspection as an optional task that can be turned off or opted out of based on config or node type. What we do with the introspection data is not within the scope of this enhancement but I can see that nodes could have an option to opt-out of any action taken based on introspection results.

@dhellmann what do you think?

Reply:
One of the key points of this work is to provide information to the user about the hardware present in the node. I don't expect we would find any reason to decide to let a node opt out of reporting that.

Reply:
Is this really only specific to baremetal users?  VMs can have PCI passthrough for dedicated hardware devices on supported platforms.  There's also virtual PCI devices.

Reply:
This enhancement is applicable to all deployments managed by "metal3". And "metal3" is currently used to manage baremetal deployments in production environments and VM based deployments only in test scenarios.